# Jupyter Notebook for Final Assignment DiSC Fall 2018

***Complete all of the prompts below. They will serve both as demonstrations of your technical/analytic skills and as source material for the substantive discussions in the remainder of your project. Download and rename this Jupyter notebook from the finalProject folder of the github, then follow the directions completely concerning each prompt.***

*All data are drawn from the [Million Song Subset](https://labrosa.ee.columbia.edu/millionsong/).*

*Please use a p<0.05 significance level for all hypothesis testing.*

## Prompt 1: Genre and Decade

***Some musical genres seem to be more enduring than others. Rock, Adult Contemporary, Country and Gospel/Christian are a few examples of genre labels that continue to be salient over many years, even if the style itself changes throughout that time. Choose 3 common genres to test whether songs (in the subset) are released with the same relative frequency across the 1980s, 1990s and 2000s between the three genres.***

In [1]:
import pandas as pd

In [17]:
#Download data for songs in genres with 50 or more songs in the subset
indata1 = pd.read_csv('https://raw.githubusercontent.com/ndporter/pythonDiSC/master/finalProject/DiSC_Million_Songs_Popular_Tags.csv')

In [18]:
###Just run this cell and don't worry about what it means unless you're interested###

#create decades variable
#function to run on each row to create a (string) decade variable
def makedecade(row):
    if 1980<=row['year']<1990:
        val='80s'
    elif 1990<=row['year']<2000:
        val='90s'
    elif 2000<=row['year']<2010:
        val='00s'
    else:
        val='missing'
    return val
#apply does the same thing to each column (axis=0) or row (axis=1) in the dataframe
#so below, I apply the makedecade function to each row to create a new column with decades
indata1['decade'] = indata1.apply(makedecade,axis=1) 

In [19]:
indata1.head()

Unnamed: 0             song_id           artist_id  \
0           1  SOCIWDW12A8C13D406  ARMJAGH1187FB546F3   
1           3  SONHOTT12A8C13493C  AR7G5I41187FB4CE6C   
2          14  SOBONFF12A6D4F84D8  ARIK43K1187B9AE54C   
3          23  SOBBUGU12A8C13E95D  ARMAC4T1187FB3FA4C   
4          30  SOGNCJP12A58A80271  ARB29H41187B98F0EF   

                 artist_name  artist_familiarity  artist_hotttnesss  \
0               The Box Tops            0.630630           0.417500   
1                   Adam Ant            0.630382           0.454231   
2              Lionel Richie            0.776676           0.553072   
3  The Dillinger Escape Plan            0.839963           0.541889   
4              Terry Callier            0.706828           0.416173   

                             title  song_hotttnesss  year   duration  \
0                        Soul Deep              NaN  1969  148.03546   
1                  Something Girls              NaN  1982  233.40363   
2          Tonight Will Be Alright              NaN  1986  307.38240   
3  Setting Fire to Sleeping Giants         0.666528  2004  207.77751   
4     Do You Finally Need A Friend         0.413740  1972  342.56934   

   loudness    tempo         artist_mbtags                 release  \
0    -9.843  121.274  classic pop and rock              Dimensions   
1    -9.013  119.293                    uk           Friend Or Foe   
2    -8.346  125.197       soul and reggae  Dancing On The Ceiling   
3    -4.264  166.862        rock and indie            Miss Machine   
4   -11.939  110.189  classic pop and rock         Occasional Rain   

   release_7digitalid   decade  
0              300822  missing  
1              287650      80s  
2               25811      80s  
3              223365      00s  
4              228326  missing

In [20]:
#remove decades outside the period of interest (and data missing on year)
indata1 = indata1[indata1['decade']!='missing']

In [21]:
indata1.head()

Unnamed: 0             song_id           artist_id  \
1           3  SONHOTT12A8C13493C  AR7G5I41187FB4CE6C   
2          14  SOBONFF12A6D4F84D8  ARIK43K1187B9AE54C   
3          23  SOBBUGU12A8C13E95D  ARMAC4T1187FB3FA4C   
7          46  SOTSEJY12AB0180F56  ART3O5Z1187B9AB043   
8          47  SOTSETT12AB01863B8  AR5U94U1187FB5CDCF   

                 artist_name  artist_familiarity  artist_hotttnesss  \
1                   Adam Ant            0.630382           0.454231   
2              Lionel Richie            0.776676           0.553072   
3  The Dillinger Escape Plan            0.839963           0.541889   
7         Stevie Ray Vaughan            0.733799           0.544258   
8               John Hammond            0.590693           0.452818   

                             title  song_hotttnesss  year   duration  \
1                  Something Girls              NaN  1982  233.40363   
2          Tonight Will Be Alright              NaN  1986  307.38240   
3  Setting Fire to Sleeping Giants         0.666528  2004  207.77751   
7                        Crossfire              NaN  1992  248.78975   
8             I Can't Be Satisfied              NaN  2002  195.31710   

   loudness    tempo    artist_mbtags                 release  \
1    -9.013  119.293               uk           Friend Or Foe   
2    -8.346  125.197  soul and reggae  Dancing On The Ceiling   
3    -4.264  166.862   rock and indie            Miss Machine   
7    -7.805  116.721   jazz and blues           Greatest Hits   
8   -10.700  123.142             folk                Bluesman   

   release_7digitalid decade  
1              287650    80s  
2               25811    80s  
3              223365    00s  
7              417945    90s  
8              575531    00s

In [23]:
#View a list of genres to choose from
indata1['artist_mbtags'].groupby(indata1['artist_mbtags']).count()

artist_mbtags
alternative rock               42
american                       93
british                        82
classic pop and rock          123
country                        45
finnish                        50
folk                           57
french                         39
german                         52
hip hop rnb and dance hall    121
jazz and blues                 36
pop and chart                  64
production music               22
punk                           44
rock                           54
rock and indie                124
soul and reggae                43
uk                            106
Name: artist_mbtags, dtype: int64

*Choose three genres to investigate and create a hypothesis about them related to the prompt above. Include a theoretical rationale justifying your hypothesis in terms of the data source and/or the genres themselves and write both the hypothesis and justification (in words) in the cell below.*

**Prompt 1 Hypothesis:**

> The number of songs released in the genres "folk", "rock" and "soul and reggae" were distributed differently across the decades of the 80s, 90s and 2000s. Of the three, soul and reggae is likely to be more closely tied to individual time periods and less consistent across the decades.

*Formally state the null (H0) hypotheses in the cell below.*

**Prompt 1 Null Hypotheses:**

> H0: Folk, rock, and soul and reggae music do not differ in their distribution across decades.

In [26]:
#Restrict to key genres by ***replacing sample with your genres of interest***
genres = ['folk','rock','soul and reggae']

def genreIndicator(row):
    temp = 0
    for genre in genres:
        if row['artist_mbtags']==genre:
            temp += 1
    return temp

indata1['keep']=indata1.apply(genreIndicator,axis=1)

indata1 = indata1[indata1['keep'] != 0]
indata1.head()

Unnamed: 0             song_id           artist_id    artist_name  \
2           14  SOBONFF12A6D4F84D8  ARIK43K1187B9AE54C  Lionel Richie   
8           47  SOTSETT12AB01863B8  AR5U94U1187FB5CDCF   John Hammond   
11          55  SODHVES12A6701CE6B  AR3793X1187FB50CB3      Chris Rea   
21          94  SORWHWY12A6702038E  ARXWAQQ1187B9AE954         Redman   
29         115  SOFRAZJ12A8C143CDB  ARR9XGB1187B9912E6    Boyz II Men   

    artist_familiarity  artist_hotttnesss                       title  \
2             0.776676           0.553072     Tonight Will Be Alright   
8             0.590693           0.452818        I Can't Be Satisfied   
11            0.643183           0.500691  Driving Home For Christmas   
21            0.743130           0.493567                 WKYA (drop)   
29            0.790923           0.601185              On Bended Knee   

    song_hotttnesss  year   duration  loudness    tempo    artist_mbtags  \
2               NaN  1986  307.38240    -8.346  125.197  soul and reggae   
8               NaN  2002  195.31710   -10.700  123.142             folk   
11         0.767728  1986  241.16200    -8.531  180.149             rock   
21         0.570106  2001  124.18567   -10.296   85.008  soul and reggae   
29         0.345802  1994  329.27302    -7.103  116.473  soul and reggae   

                    release  release_7digitalid decade  keep  
2    Dancing On The Ceiling               25811    80s     1  
8                  Bluesman              575531    00s     1  
11  The Road To Hell Part 2               17220    80s     1  
21              Malpractice                8552    00s     1  
29                   Legacy              523481    90s     1

*Test your hypotheses using the appropriate statistical test. You do not need to worry about missing data here.*

In [27]:
#run a crosstab table of genre and decade
tab = pd.crosstab(indata1['decade'],indata1['artist_mbtags'])
tab
#it looks like soul and reggae may be different (relatively more common in 2000s than before)

artist_mbtags  folk  rock  soul and reggae
decade                                    
00s              31    33               32
80s               7     6                3
90s              19    15                8

In [29]:
#test hypothesis
from scipy.stats.contingency import chi2_contingency
chi2_contingency(tab)

(4.26062066503468,
 0.37188218379120147,
 4,
 array([[35.53246753, 33.66233766, 26.80519481],
        [ 5.92207792,  5.61038961,  4.46753247],
        [15.54545455, 14.72727273, 11.72727273]]))

*Report the results of your null hypothesis test formally. What test did you use and what was the significance (p) value? Did you reject or fail to reject the null hypothesis?*

**Prompt 1 Null Hypothesis Findings:**

> Using a chi-squared test, I find a significance of p<0.37 and fail to reject the null hypothesis.

*Interpret the results of your null hypothesis test in words. What do your findings tell us about these genres (at least in the Million Song Data Subset)?*

> There is not sufficient evidence to find a difference in the distribution of number of folk, rock and soul & reggae music across the 80s, 90s and 2000s.

*If you rejected the null hypothesis: describe one possible mechanism that might drive this difference (e.g. why would the distribution in the data across decades differ by genre).*

*If you failed to reject the null hypothesis: describe one possible reason for the null finding (e.g. why wouldn't the distribution in the data across decades differ by genre).*

*In either case, your explanation may include any combination of substantive, epistemological, or measurement issues.*

**Prompt 1 Discussion of Findings:**

> The small sample of songs with each tag may result in reduced statistical power to detect real differences. The crosstab and expected values appeared to differ, so it may just be an issue of incomplete data. This could have resulted from the combination of sampling songs for the Million Song Subset (1/100th of the original sample) and the restriction to only the first artist genre tag. Also, the analysis is at the song level while the tags are at the artist level, which may also reduce power.

## Prompt 2: Artist and Musical Characteristics

***Investigate whether a musical characteristic (loudness, tempo, or duration) varies between two artists of your choice.***

In [1]:
import pandas as pd

In [2]:
#Download data for artists with 10 or more songs in the subset
indata2 = pd.read_csv('https://raw.githubusercontent.com/ndporter/pythonDiSC/master/finalProject/DiSC_Million_Songs_Artists.csv')

In [3]:
indata2.head()

Unnamed: 0  Unnamed: 0.1             song_id           artist_id  \
0          71            71  SOVJXVJ12A8C13517D  ARCCRTI11F4C845308   
1          76            76  SOWHYUG12A58A77FAB  ARODBRG1187FB3FD99   
2         270           270  SOTVEKK12A8C1392E1  ARX9YIP1187B98A656   
3         344           344  SOUPIRU12A6D4FA1E1  ARJIE2Y1187B994AB7   
4         406           406  SOGXQRO12A8AE47D50  ARN03F71187FB4E3F4   

                 artist_name  artist_familiarity  artist_hotttnesss  \
0                White Heart            0.477365           0.337507   
1  Nick Cave & The Bad Seeds            0.694022           0.479408   
2                    RUN-DMC            0.686989           0.606630   
3                Line Renaud            0.470212           0.320087   
4                        Ayo            0.522590           0.390641   

                                               title  song_hotttnesss  year  \
0  Where The Thunder Roars (Tales Of Wonder Album...         0.265861     0   
1              She Fell Away (2009 Digital Remaster)              NaN     0   
2                                         Hard Times              NaN  1984   
3                                Der Kleine Dompfaff              NaN     0   
4                                           Complain         0.000000  2006   

    duration  loudness    tempo    artist_mbtags  \
0  298.84036    -9.998  113.277             none   
1  269.26975   -11.751  144.603        seen live   
2  234.65751   -12.750  119.670          hip-hop   
3  152.92036    -9.843  124.750             none   
4  212.71465   -11.260  128.689  soul and reggae   

                                            release  release_7digitalid  
0                                   Tales Of Wonder              148547  
1  Your Funeral... My Trial (2009 Digital Remaster)              413222  
2                                           Run DMC              293884  
3                              Platinum Line Renaud              272703  
4                                            Joyful              350343

In [5]:
#remove missing data on all variables
vars = ['loudness','tempo','duration']
indata2 = indata2.dropna(subset=vars) #defined Nan
for var in vars:
    indata2 = indata2[indata2[var]!=0] #missing set to zero

In [6]:
#Display all artists to choose from
indata2['artist_name'].groupby(indata2['artist_name']).count()

artist_name
Aerosmith                                12
Ayo                                      10
Bill & Gloria Gaither                    11
Jacques Douai                            10
Line Renaud                              11
Mario Rosenstock                         13
Michael Jackson                          10
Mikey Dread                              11
Nick Cave & The Bad Seeds                11
Phil Collins                             12
Pumpkin Buzzard                          10
RUN-DMC                                  10
Radiohead                                11
Red Hot Chili Peppers                    10
Selena                                   10
Shakira                                  10
Snow Patrol                              11
Stevie Ray Vaughan And Double Trouble    10
Sugar Minott                             12
The Del McCoury Band                     10
The Jackson Southernaires                12
White Heart                              10
Name: artist_name, d

*Choose two artists to investigate and create a hypothesis about them related to the prompt above. Include a rationale justifying your hypothesis in terms of the data source and/or the artists themselves and write both the hypothesis and justification (in words) in the cell below.*

**Prompt 2 Hypothesis:**

> I hypothesize that RUN-DMC's music is louder than Bill and Gloria Gaither's because Hip-Hop tends to be louder (or at least played louder) than Gospel.

*Formally state the null (H0) hypotheses in the cell below.*

**Prompt 2 Null Hypotheses:**

> RUN-DMC does not have higher mean loudness than Bill and Gloria Gaither.

*Test your hypotheses using the appropriate statistical test. You do not need to worry about missing data here.*

In [10]:
#create subsets to compare
gaither = indata2[indata2['artist_name']=='Bill & Gloria Gaither']
run = indata2[indata2['artist_name']=='RUN-DMC']
print('gaither mean:',gaither['loudness'].mean())
print('run-dmc mean:',run['loudness'].mean())

gaither mean: -13.38881818181818
run-dmc mean: -9.5013


In [11]:
#at least in the sample, my hypothesis appears to be correct; the mean loudness for run is higher
#let's see if we have the statistical power to draw a conclusion
#Test the difference of means using a T-test
import statsmodels.stats.weightstats as stats
stats.ttest_ind(gaither['loudness'],run['loudness'])

(-2.6179667186158575, 0.016923139260728696, 19.0)

*Report the results of your null hypothesis test formally. What test did you use and what was the significance (p) value? Did you reject or fail to reject the null hypothesis?*

**Prompt 2 Null Hypothesis Findings:**

> Using a t-test of the difference of means, I reject the null hypothesis with a p-value of 0.017.

*Interpret the results of your null hypothesis test in words. What do your findings tell us about these artists (at least in the Million Song Data Subset)?*

> I find a significant difference in mean loudness between RUN-DMC and Bill and Gloria Gaither, with RUN-DMC being louder. At least for these two artists, the patterns of loudness of recordings tends to mirror stereotypical patterns of preference for loudness, with hip-hop being louder than gospel.

*If you rejected the null hypothesis: describe one possible mechanism that might drive this difference (e.g. why would the variable's values differ between the two artists).*

*If you failed to reject the null hypothesis: describe one possible reason for the null finding (e.g. why wouldn't the variable's values differ between the two artists).*

*In either case, your explanation may include any combination of substantive, epistemological, or measurement issues.*

**Prompt 2 Discussion of Findings:**

> This seems unlikely to be an artifact of measurement, given that the variable was created automatically based on measured volume in the songs. Instead, it probably reflects the differing preferences of fans of different genres of music and/or the differences in the capabilities of the equipment they tend to use for playback.

# Prompt 3: Year and Hotttnesss/Familiarity

***There has been some speculation in this class that the hotttnesss and familiarity measurement in the dataset might be biased toward music that was newer  when the dataset was collected (2010). Investigate the relationship between year (not recoded) and the three variables to test whether there is reason to further investigate the potential for such bias.***

In [33]:
import pandas as pd

In [34]:
#Download data for the full subset
indata3 = pd.read_csv('https://raw.githubusercontent.com/ndporter/pythonDiSC/master/finalProject/DiSC_Million_Songs_Data.csv')

In [39]:
indata3.head()

year  song_hotttnesss  artist_hotttnesss  artist_familiarity
4   2007         0.604501           0.401724            0.651046
8   1984         0.265861           0.332276            0.426668
15  1987         0.405116           0.447935            0.636424
18  2004         0.684136           0.513463            0.707200
23  2004         0.666528           0.541889            0.839963

In [38]:
#remove missing data on all variables
vars = ['year','song_hotttnesss','artist_hotttnesss','artist_familiarity']
indata3 = indata3.dropna(subset=vars) #defined NaN
for var in vars:
    indata3 = indata3[indata3[var]!=0] #missing set to zero
indata3 = indata3[vars] #drop exra variables

*Choose one variable (song_hotttnesss, artist_hotttnesss or artist_familiarity) to investigate and create a hypothesis about its relationship to year based on the prompt above. Include a rationale justifying your hypothesis in terms of the data source and/or the artists themselves and write both the hypothesis and justification (in words) in the cell below.*

**Prompt 3 Hypothesis:**

> I hypothesize that song_hotttnesss will have a positive association with year because more recent music is more likely to be considered hot in whatever metric last.fm used.

*Formally state the null (H0) hypotheses in the cell below.*

**Prompt 3 Null Hypotheses:**

> H0: Song_hotttnesss does not have a positive association with year

*Create a correlation matrix of all variables (note that year is in calendar years, not decades like in prompt 1)*

In [40]:
#Use pandas to get correlation matrix for all variables
corrMatr = indata3.corr()
corrMatr

year  song_hotttnesss  artist_hotttnesss  \
year                1.000000         0.065018           0.071455   
song_hotttnesss     0.065018         1.000000           0.459118   
artist_hotttnesss   0.071455         0.459118           1.000000   
artist_familiarity  0.102442         0.499423           0.782037   

                    artist_familiarity  
year                          0.102442  
song_hotttnesss               0.499423  
artist_hotttnesss             0.782037  
artist_familiarity            1.000000

*Interpret the sign and relative sizes of the correlations of year with all three other variables. You do not need to discuss significance yet.* 

> All three song and artist variables are positively correlated with year, although the correlations are relatively small, the largest being 0.102 for artist_familiarity and year. Thus more recent music (and the artists that published it) tend to be rated more highly on both hotttnesss and familiarity.

*Test your hypothesis using the appropriate statistical test. You do not need to worry about missing data here.*

In [44]:
#use scipy to get p-value for hypothesis testing
import scipy.stats  as stats
stats.pearsonr(indata3['year'], indata3['song_hotttnesss'])

(0.06501756123228783, 0.0007270482408098302)

*Report the results of your null hypothesis test formally. What test did you use and what was the significance (p) value? Did you reject or fail to reject the null hypothesis?*

**Prompt 3 Null Hypothesis Findings:**

> I use Pearson's correlation and reject the null hypothesis (p<0.0007).

*Interpret the results of your null hypothesis test in words. What do your findings tell us about the variables and/or data?*

> I conclude that there is a positive association between year and song_hotttnesss.

*If you rejected the null hypothesis: describe one possible mechanism that might drive this difference (e.g. why would the variable's values in the data differ across years).*

*If you failed to reject the null hypothesis: describe one possible reason for the null finding (e.g. why wouldn't the variable's values in the data differ across years).*

*In either case, your explanation may include any combination of substantive, epistemological, or measurement issues.*

**Prompt 3 Discussion of Findings:**

> I suspect that, like the prompt suggested, there is a recency bias in the ratings contained in the sources of the million song data. Based on this, I would recommmend that others interested in using the data are cautious about interpreting the hotttnesss and familiarity measures as indicative of anything broader than the situation when the data were collected in 2010.